modules/tinymce/src/plugins/table/test/ts/browser/InvalidColCountTest.ts

```diff
@@ -0,0 +1,129 @@
+import { UiFinder } from '@ephox/agar';
+import { context, describe, it } from '@ephox/bedrock-client';
+import { Arr, Type } from '@ephox/katamari';
+import { TinyAssertions, TinyDom, TinyHooks, TinySelections } from '@ephox/wrap-mcagar';
+import { assert } from 'chai';
+
+import Editor from 'tinymce/core/api/Editor';
+import Plugin from 'tinymce/plugins/table/Plugin';
+import Theme from 'tinymce/themes/silver/Theme';
+
+describe('browser.tinymce.plugins.table.InvalidColCountTest', () => {
+  const hook = TinyHooks.bddSetupLight<Editor>({
+    plugins: 'table',
+    base_url: '/project/tinymce/js/tinymce',
+  }, [ Plugin, Theme ], true);
+
+  const assertBasicTablePresence = (editor: Editor, tdCount: number, colCount?: number) => {
+    TinyAssertions.assertContentPresence(editor, {
+      td: tdCount,
+    });
+
+    if (Type.isNonNullable(colCount)) {
+      const cols = UiFinder.findAllIn(TinyDom.body(editor), 'col');
+      assert.lengthOf(cols, colCount);
+    }
+  };
+
+  // Note: The goal of these tests is to make sure an exception is not thrown and each table operation works
+  // when there are less or more col elements than actual columns
+  Arr.each([
+    {
+      label: 'less cols',
+      colgroupHtml: '<colgroup><col /></colgroup>',
+      assertColCount: true
+    },
+    {
+      label: 'less cols with span',
+      colgroupHtml: '<colgroup><col span="2" /></colgroup>',
+    },
+    {
+      label: 'more cols',
+      colgroupHtml: '<colgroup><col /><col /><col /><col /></colgroup>',
+      assertColCount: true
+    },
+    {
+      label: 'more cols with span at beginning',
+      colgroupHtml: '<colgroup><col span="3" /><col /></colgroup>',
+    },
+    {
+      label: 'more cols with span at end',
+      colgroupHtml: '<colgroup><col /><col span="3" /></colgroup>',
+    },
+    {
+      label: 'single col with span greater than cells',
+      colgroupHtml: '<colgroup><col span="4" /></colgroup>',
+    }
+  ], (scenario) => {
+    context(scenario.label, () => {
+      const tableHtml = '<table>' +
+        scenario.colgroupHtml +
+        '<tbody>' +
+        '<tr><td>1</td><td>2</td><td>3</td></tr>' +
+        '<tr><td>4</td><td>5</td><td>6</td></tr>' +
+        '</tbody>' +
+        '</table>';
+
+      const assertCount = scenario.assertColCount === true;
+      const countAfterInsertColumn = assertCount ? 4 : undefined;
+      const countAfterDeleteColumn = assertCount ? 2 : undefined;
+      const countAfterNonColumnAction = assertCount ? 3 : undefined;
```

Comment:
Instead of using undefined for everything here, can we please use `Optional` instead here, as that'd make more sense with how it's used as well. e.g
```suggestion
      const countAfterInsertColumn = Optionals.someIf(assertCount, 4);
      const countAfterDeleteColumn = Optionals.someIf(assertCount, 2);
      const countAfterNonColumnAction = Optionals.someIf(assertCount, 3);
```

Reply:
Done